### PR TITLE
NodeJS tutorial 

### PR DESCRIPTION
**Pull request by ash-sxn**
- I've rebased the python tutorial's branch in this one because the init and teardown scripts were already updated in the Python tutorial.

### CODE REVIEW

**Review by gounthar — COMMENTED** [review]
The port 3000 is also open.

**Review by gounthar — CHANGES_REQUESTED** [review]
The `3000` port is still opened.

**Comment by ash-sxn** [comment]
> The `3000` port is still open.

It's because the tutorial needs to use port 3000 for viewing the website 
check it out [here](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm/#add-a-final-deliver-stage-to-your-pipeline)

**Comment by gounthar** [comment]
Fine with me then.
Thanks!

**Review by gounthar — APPROVED** [review]

File: 05_nodejs/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+## Example 5
+###  NodeJS Tutorial
+This example uses 02 example with custom ssh agent to perform [NodeJS Tutorial](https://www.jenkins.io/doc/tutorials/simple-node-js-react-npm-app /) 
+### Files 
+* `dockerfiles/Dockerfile` is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
```
Comment (gounthar) [inline 05_nodejs/README.md:5]:
```suggestion

* `dockerfiles/Dockerfile` is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
```

File: 05_nodejs/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+* `dockerfile/agent/Dockerfile` is used to create a custom agent image from `jenkins/ssh-agent` with nodeJS installed to perform the tutorial
```
Comment (gounthar) [inline 05_nodejs/README.md:6]:
```suggestion
* `The dockerfile/agent/Dockerfile` is used to create a custom agent image from `jenkins/ssh-agent` with nodeJS installed , enabling it to perform the tutorial.
```

File: 05_nodejs/README.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+* `keygen.sh` is a script that generates new ssh-keys in secrets directory everytime it runs and updates the docker compose file with new keys
```
Comment (gounthar) [inline 05_nodejs/README.md:7]:
```suggestion
* `keygen.sh` is a script that generates new `SSH` keys in the secrets directory every time it runs and updates the Docker Compose file with the new keys.
```

File: 05_nodejs/README.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
+## Example 5
+###  NodeJS Tutorial
+This example uses 02 example with custom ssh agent to perform [NodeJS Tutorial](https://www.jenkins.io/doc/tutorials/simple-node-js-react-npm-app /) 
+### Files 
+* `dockerfiles/Dockerfile` is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
+* `dockerfile/agent/Dockerfile` is used to create a custom agent image from `jenkins/ssh-agent` with nodeJS installed to perform the tutorial
+* `keygen.sh` is a script that generates new ssh-keys in secrets directory everytime it runs and updates the docker compose file with new keys 
+* `fork-and-clone.sh` is script that takes github credentials from user and fork and clone a repo used in the tutorial from a docker container using Dockerfile
```
Comment (gounthar) [inline 05_nodejs/README.md:8]:
```suggestion
* `fork-and-clone.sh` is a script that takes GitHub credentials from the user, forks, and clones a repository used in the tutorial from within a Docker container using the Dockerfile.
```

File: 05_nodejs/README.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
+## Example 5
+###  NodeJS Tutorial
+This example uses 02 example with custom ssh agent to perform [NodeJS Tutorial](https://www.jenkins.io/doc/tutorials/simple-node-js-react-npm-app /) 
+### Files 
+* `dockerfiles/Dockerfile` is used to create a custom image from the `jenkins/jenkins` image and copies `plugins.txt` and `jenkins.yaml` into the container for plugins installation and JCasC configuration.
+* `dockerfile/agent/Dockerfile` is used to create a custom agent image from `jenkins/ssh-agent` with nodeJS installed to perform the tutorial
+* `keygen.sh` is a script that generates new ssh-keys in secrets directory everytime it runs and updates the docker compose file with new keys 
+* `fork-and-clone.sh` is script that takes github credentials from user and fork and clone a repo used in the tutorial from a docker container using Dockerfile 
+* ` jenkins.yaml` is a JCasC configuration file that can be edited to customize the build accordingly.
+* `gitpodURl.sh` is a script that replace `http://127.0.0.1:8080` with the gitpod appropriate url in the jenkins.yaml file
```
Comment (gounthar) [inline 05_nodejs/README.md:10]:
```suggestion
* `gitpodURl.sh` is a script that replaces `http://127.0.0.1:8080` with the appropriate Gitpod URLin the `jenkins.yaml` file
```

File: 05_nodejs/README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+* `plugins.txt` is a list of plugins that are installed with the help of [`jenkins-plugin-cli`](https://www.jenkins.io/doc/book/managing/plugins/#install-with-cli).
+* The `docker-compose.yaml` file defines two containers: the Jenkins controller and an agent. It then connects them to a common network, sets the private key in the controller and the public key in the agent, defines a named volume, and builds the image from the `Dockerfile`.
+
+### Steps to run the example 
+* Make sure you are in the `05_nodejs` directory by using the command `pwd`.
```
Comment (gounthar) [inline 05_nodejs/README.md:15]:
```suggestion

* Make sure you are in the `05_nodejs` directory by using the command `pwd`. 
```

File: 05_nodejs/README.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+* `keygen.sh` is a script that generates new ssh-keys in secrets directory everytime it runs and updates the docker compose file with new keys 
+* `fork-and-clone.sh` is script that takes github credentials from user and fork and clone a repo used in the tutorial from a docker container using Dockerfile 
+* ` jenkins.yaml` is a JCasC configuration file that can be edited to customize the build accordingly.
+* `gitpodURl.sh` is a script that replace `http://127.0.0.1:8080` with the gitpod appropriate url in the jenkins.yaml file
+* `plugins.txt` is a list of plugins that are installed with the help of [`jenkins-plugin-cli`](https://www.jenkins.io/doc/book/managing/plugins/#install-with-cli).
+* The `docker-compose.yaml` file defines two containers: the Jenkins controller and an agent. It then connects them to a common network, sets the private key in the controller and the public key in the agent, defines a named volume, and builds the image from the `Dockerfile`.
+
+### Steps to run the example 
+* Make sure you are in the `05_nodejs` directory by using the command `pwd`. 
+* Run the script with `./keygen.sh` command to generate ssh keys
```
Comment (gounthar) [inline 05_nodejs/README.md:16]:
```suggestion
* Run the script with the `./keygen.sh` command to generate new `SSH` keys 
```

File: 05_nodejs/README.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+* ` jenkins.yaml` is a JCasC configuration file that can be edited to customize the build accordingly.
+* `gitpodURl.sh` is a script that replace `http://127.0.0.1:8080` with the gitpod appropriate url in the jenkins.yaml file
+* `plugins.txt` is a list of plugins that are installed with the help of [`jenkins-plugin-cli`](https://www.jenkins.io/doc/book/managing/plugins/#install-with-cli).
+* The `docker-compose.yaml` file defines two containers: the Jenkins controller and an agent. It then connects them to a common network, sets the private key in the controller and the public key in the agent, defines a named volume, and builds the image from the `Dockerfile`.
+
+### Steps to run the example 
+* Make sure you are in the `05_nodejs` directory by using the command `pwd`. 
+* Run the script with `./keygen.sh` command to generate ssh keys 
+* Then use the command `docker compose up -d --build` to run the example.
+* after the containers are successfully running (check with `docker compose ps`). You can access the controller on [`http://localhost:8080`](http://localhost:8080)
```
Comment (gounthar) [inline 05_nodejs/README.md:18]:
```suggestion
* After the containers are successfully running (check with `docker-compose ps`), you can access the controller by navigating to [`http://localhost:8080/`](http://localhost:8080/).
```

File: 05_nodejs/README.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+* You can log in using the `admin` username and the `admin` password.
+### Steps to follow the [NodeJS Tutorial](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm)
```
Comment (gounthar) [inline 05_nodejs/README.md:20]:
```suggestion

### Steps to follow the [NodeJS Tutorial](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm)
```

File: 05_nodejs/README.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+- [Fork and clone the sample repository](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm/#fork-and-clone-the-sample-repository-on-github)
+
+  - Make sure you are in `05_nodejs` directory and Dockerfile and fork-and-clone.sh is present there
```
Comment (gounthar) [inline 05_nodejs/README.md:24]:
```suggestion
  - Before proceeding, ensure that you are in the `05_nodejs` directory, and confirm that both the `Dockerfile` and `fork-and-clone.sh` are present in this directory.
```

File: 05_nodejs/README.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  - Run `./fork-and-clone.sh` command, and enter your github credentails to get a clone of a your fork of [`simple-node-js-react-npm-app `](https://github.com/jenkins-docs/simple-node-js-react-npm-app )  
+  - If everything ran sucessfully you should have a `simple-node-js-react-npm-app ` directory 
+- [Create your initial Pipeline as a Jenkinsfile](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm/#create-your-pipeline-project-in-jenkins)
+  
+  - get inside the `simple-node-js-react-npm-app ` with `cd simple-node-js-react-npm-app ` and create a `Jenkinsfile` with this content
```
Comment (gounthar) [inline 05_nodejs/README.md:29]:
```suggestion
  - Go inside the `simple-node-js-react-npm-app` directory by using the command `cd simple-node-js-react-npm-app`.
  - Once inside the directory, create a `Jenkinsfile` and add the specified content to it:
```

File: 05_nodejs/README.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        }
+    }
+  }
+  
+  ```
+  - Save your `Jenkinsfile` 
+  - run the commands `git add .` to add the Jenkinsfile to the staging then run `git commit -m "Add initial Jenkinsfile"` to commit the changes, then run `git push origin master` to push the changes to you forked repo  
+  - Go to Jenkins, log in and click create new jobs under Welcome to Jenkins!
+  - In the Enter an item name field, specify the name for your new Pipeline project (e.g. simple-node-js-react-npm-app ).
```
Comment (gounthar) [inline 05_nodejs/README.md:46]:
```suggestion
  - Save your `Jenkinsfile`.
  - Run the commands `git add .` to add the `Jenkinsfile` to the staging, then run `git commit -m "Add initial Jenkinsfile"` to commit the changes. Finally, run `git push origin master` to push the changes to your forked repo.
 - Go to Jenkins, login, and click "Create new jobs" under "Welcome to Jenkins!".
  - In the "Enter an item name" field, specify the name for your new Pipeline project (e.g., "`simple-node-js-react-npm-app`").
```

File: 05_nodejs/README.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+  - Click the Pipeline tab at the top of the page to scroll down to the Pipeline section.
+
+  - From the Definition field, choose the Pipeline script from SCM option. This option instructs Jenkins to obtain your Pipeline from Source Control Management (SCM), which will be your forked Git repository.
+
+  - From the SCM field, choose Git.
+
+  - In the Repository URL field, specify the directory path of your forked repository above, which is https://github.com/<username>/simple-node-js-react-npm-app.git (replace <username> with your github username)
+
+  - Click Save to save your new Pipeline project. You’re now ready to begin. Click `Build Now` button to build your project
```
Comment (gounthar) [inline 05_nodejs/README.md:58]:
```suggestion
  - Scroll down and click "Pipeline", then click "OK" at the end of the page.
  - Click the "Pipeline" tab at the top of the page to scroll down to the Pipeline section.
  - In the "Definition" field, choose the "Pipeline script from SCM" option. This option instructs Jenkins to obtain your Pipeline from Source Control Management (SCM), which will be your forked Git repository.
  - From the "SCM" field, choose "Git".
  - In the "Repository URL" field, specify the directory path of your forked repository above, which is `https://github.com/<username>/simple-node-js-react-npm-app.git` (replace `<username>` with your GitHub username).
  - Click "Save" to save your new Pipeline project. You’re now ready to begin. Click the "Build Now" button to build your project.
```

File: 05_nodejs/README.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+- [Add a test stage to your Pipeline](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm/#add-a-test-stage-to-your-pipeline)
+
+  - Again open your Jenkinsfile and Copy and paste the following Declarative Pipeline syntax under the Build stage of your Jenkinsfile and save the file
```
Comment (gounthar) [inline 05_nodejs/README.md:62]:
```suggestion
  - Once again, open your `Jenkinsfile` and copy and paste the following Declarative Pipeline syntax under the Build stage of your Jenkinsfile. Then, save the file.
```

File: 05_nodejs/README.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+        stage('Test') {
+            steps {
+                sh './jenkins/scripts/test.sh'
+            }
+        }
+  - run the commands `git add .` to add the edited Jenkinsfile to the staging then run `git commit -m "Add 'Test' stage"` to commit the changes, then run `git push origin master` to push the changes to you forked repo   
+  - The Press the `Build Now` button again to run the both Build and Test stage
+- [Add a final deliver stage to your Pipeline](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm/#add-a-final-deliver-stage-to-your-pipeline)
```
Comment (gounthar) [inline 05_nodejs/README.md:71]:
```suggestion
  - Run the command `git add .` to add the edited `Jenkinsfile` to the staging, then run `git commit -m "Add 'Test' stage"` to commit the changes. Finally, run `git push origin master` to push the changes to your forked repo.
  - Press the "Build Now" button again to run both the Build and Test stages.
 - [Add a final deliver stage to your Pipeline](https://www.jenkins.io/doc/tutorials/build-a-node-js-and-react-app-with-npm/#add-a-final-deliver-stage-to-your-pipeline)
```

File: 05_nodejs/README.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  
+  - Again open your Jenkinsfile and Copy and paste the following Declarative Pipeline syntax under the Test stage of your Jenkinsfile and save the file
```
Comment (gounthar) [inline 05_nodejs/README.md:73]:
```suggestion
  - Again, open your `Jenkinsfile` and copy-paste the following Declarative Pipeline syntax under the Test stage of your Jenkinsfile. Then, save the file.
```

File: 05_nodejs/README.md [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+        stage('Deliver') { 
+            steps {
+                sh './jenkins/scripts/deliver.sh' 
+                input message: 'Finished using the website? (Click "Proceed" to continue)' 
+                sh './jenkins/scripts/kill.sh' 
+            }
+        }
+  - run the commands `git add .` to add the edited Jenkinsfile to the staging then run `git commit -m "Add 'Deliver' stage"` to commit the changes, then run `git push origin master` to push the changes to you forked repo   
+  - The Press the `Build Now` button again to run Build, Test and Deliver stages
```
Comment (gounthar) [inline 05_nodejs/README.md:83]:
```suggestion
  - Run the command `git add .` to add the edited `Jenkinsfile` to the staging, then run `git commit -m "Add 'Deliver' stage"` to commit the changes. Afterward, run `git push origin master` to push the changes to your forked repository.
  - Press the "Build Now" button again to run the Build, Test, and Deliver stages.
```

File: 05_nodejs/README.md [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+### Clean Up Instructions 
+
+* To stop and clean up a working Docker Compose setup, you can use the following command:
+`docker compose down` 
+
+This command stops the containers defined in your Docker Compose file and removes them along with any networks, volumes, and other resources created by the Compose setup.
+It effectively cleans up the resources associated with your Docker Compose environment.
```
Comment (gounthar) [inline 05_nodejs/README.md:91]:
```suggestion
This command stops the containers defined in your Docker Compose file and removes them, along with any networks, volumes, and other resources created by the Compose setup.
It effectively cleans up the resources associated with your Docker Compose environment.
```

File: 05_nodejs/README.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+By default, `docker compose down` removes the stopped containers and networks, but it preserves named volumes.
+If you also want to remove the named volumes, you can use the `--volumes` flag:
+`docker compose down --volumes` 
+
+This ensures that all resources, including volumes, are removed.
```
Comment (gounthar) [inline 05_nodejs/README.md:97]:
```suggestion
This command will ensure that all associated resources, including named volumes, are removed when bringing down the Docker Compose environment.
```

**Review by gounthar — CHANGES_REQUESTED** [review]

**Review by gounthar — COMMENTED** [review]
Thanks.
Now, what about adding the GitHub Action for Node?

**Review by gounthar — APPROVED** [review]
Ok, I've seen the GitHub Action is part of #122 .

**Pull request merged**